src/mongoose_acc.erl

```diff
@@ -1,300 +1,263 @@
 %%%-------------------------------------------------------------------
-%%% @doc
-%%% This module encapsulates a data type which will is instantiated when stanza
-%%% enters the system and is passed all the way along processing chain.
-%%% Its interface is map-like, and you can put there whatever you want, bearing in mind two things:
-%%% 1. It is read-only, you can't change value once you wrote it (accumulator is to accumulate)
-%%% 2. Whatever you put will be removed before the acc is sent to another c2s process
+%%% File    : mongoose_acc.erl
+%%% Author  : Piotr Nosek <piotr.nosek@erlang-solutions.com>
+%%% Author  : Bartlomiej Gorny <bartlomiej.gorny@erlang-solutions.com>
+%%% Purpose : Mongoose accumulator implementation
+%%% Created : 11 Sep 2018 by Piotr Nosek <piotr.nosek@erlang-solutions.com>
 %%%
-%%% There are three caveats to the above:
-%%% 1. Although you can not put to an existing key, you can append as many times as you like
-%%% 2. A special key 'result' is writeable and is meant to be used to get return value from hook calls
-%%% 3. If you want to pass something to another c2s process you can use add_prop/3 - values
-%%%    put there are not stripped.
-%%% @end
+%%% NS:Key conventions:
+%%% * hook:result should be used to return hook processing result
+%%% * iq:* contains useful IQ metadata but must be provided by mongoose_iq.erl
 %%%-------------------------------------------------------------------
 -module(mongoose_acc).
 -author("bartek").
+-author("piotr.nosek@erlang-solutions.com").
 
 -include("jlib.hrl").
 -include("mongoose.hrl").
 
 %% API
--export([new/0, from_kv/2, put/3, get/2, get/3, find/2, append/3, remove/2]).
--export([increment/2, get_counter/2]).
--export([add_prop/3, get_prop/2]).
--export([from_element/1, from_map/1, update_element/4, update/2, is_acc/1, require/2]).
--export([strip/1, strip/2, record_sending/4, record_sending/6]).
--export([dump/1, to_binary/1]).
--export([to_element/1]).
--export_type([t/0]).
--export([from_element/3]).
+% Constructor
+-export([new/1]).
+% Access to built-in fields
+-export([
+         ref/1,
+         timestamp/1,
+         lserver/1,
+         element/1,
+         to_jid/1,
+         from_jid/1,
+         stanza_name/1,
+         stanza_type/1,
+         stanza_ref/1
+        ]).
+% Stanza update
+-export([update_stanza/2]).
+% Access to namespaced fields
+-export([
+         set/4,
+         set/5,
+         append/4,
+         append/5,
+         get/3,
+         get/4,
+         delete/3
+        ]).
+% Strip and replace stanza
+-export([strip/2]).
+%% Debug API
+-export([dump/1]).
+
+-type location() :: {Module :: module(), Function :: atom(), Line :: pos_integer()}.
+-type stanza_metadata() :: #{
+        element := exml:element(),
+        from_jid := jid:jid(),
+        to_jid := jid:jid(),
+        name := binary(),
+        type := binary(),
+        ref := reference()
+       }.
 
 %% if it is defined as -opaque then dialyzer fails
```

Comment:
I see this was an old comment. Would it still fail with the new structure and types?

Reply:
Don't know, will check.